### ISSUE TITLE
Ability to comment the source code

### ISSUE THREAD

**Issue by fracz** [JBirdVegas/external_jbirdvegas_mGerrit#46]
I have wrote it in a comment at Google Play but this seems to be more suitable place. It would be great if we could comment source code (I mean adding comments to lines of code). Now, the only possiblity (I'm aware of) is to add an overall comment for the change.

**Comment by p4r4n01d** [comment]
Yes, this is a great idea and is definitely something on the list for the future direction of authenticated actions.

However, since this is a large change and I am not sure of the UI/UX involved on how best to achieve this, it currently has a low priority. There are a number of other authenticated actions planned sooner than this which there is currently no workaround possible in mGerrit (not including going to the web interface).

Some notes:
- The diff viewer will need to provide more context to diffs, with the intention of replicating something functionally similar to either the side-by-side view or the unified view available on the web interface. There will probably be architectural changes required here and needs more thought
- The UI/UX needs more thought into how to make doing this practical on a small screen. There is more flexibility on a 10" tablet where we could fit a side-by-side view than a 4"/4.5" phone where there barely is enough space to fit the diff itself. This will need to be considered as we do support phones and tablets (I do nearly all of my dev on a phone anyway).

I hope this doesn't come across as I am shooting your idea down - this feature would be amazing and is definitely the direction I want to see mGerrit heading towards - it just needs plenty more thought and help on it. Any thoughts on this @JBirdVegas?

**Comment by fracz** [comment]
Well, I have some experience with this as I was interested in mobile code reviews some time. I have created [a general-purpose applicaiton for mobile code reviews](https://github.com/fracz/mcr) which then was used to launch a case study to check if code reviews on mobile devices make sense. In fact, the data I have collected has shown that mobile code reviews are as afficient as code reviews on large PC screens (you can find experiment details, collected data and conclusions in my paper: [Can the source code be reviewed on a smartphone?](https://www.researchgate.net/publication/282653909_Can_the_source_code_be_reviewed_on_a_smartphone)). The application does not use Gerrit in any way. It is designed to open source code from SD card. But it has some _great_ features like recording voice comments for example.

Moreover, as I'm currently working at a university, I had students that made [an application that uses Gerrit API](https://github.com/mobilecodereviewer/MobileCodeReviewer) and provides similar functionality as mGerrit. Unfortunately, it was never released in Google Play and the project was shut down immediately after they have finished their thesis. It had some stability issues and stopped working with Gerrit 2.11 and above.

As for the UX involved in both applications, they:
- allow to display code in two modes - unified diff (as mgerrit) and the new version only; we decided there is no place for side by side diff on small screens
- allow to add a comment to the line of _new version_ of code when you tapped the line
- display a modal window with textarea when you can enter and save the comment for the tapped line
- highlight the line number that has comments with some fancy color
- display the comments added to a line after you tap it, in the same modal, above the textarea

I can provide some screenshots if you wish.

**Comment by JBirdVegas** [comment]
I'd love to see some screenshots.  I really like the idea.  The diff UI could probably use some love from a UX designer.  Problem become worse when considering how comments on source should be shown.  For example if you allow commenting then you must also allow threads of comments at source locations, Blah.java line 43 has 4 comments.  

I'll have to double check but right now I think we are ignoring the inline comments when getting the source.  The only comments we are displaying are the comments on the review not on the diff.  Great idea.

**Comment by p4r4n01d** [comment]
To clarify, we are not fetching the inline comments at all and there would need to be some database modifications to properly do so. The only comments we are displaying are the comments on the change as a whole rather than those on a diff. Note that comments from all the patchsets (revisions) are displayed, while we only display the diff from the current patchset.

@fracz Your paper looks very interesting. I have started to read through it and I hope you don't mind if I share (link) it with others at work.

**Comment by fracz** [comment]
As for the first app I have described (no Gerrit-integration) - it was able to display the current version of the file only.

![](https://www.researchgate.net/profile/Wojciech_Fracz/publication/282653909/figure/fig1/AS:282274554761217@1444310906029/Figure-1-Source-code-displayed-in-mobile-application.png)

And the features are as follows:

1. Predefined comments (one-click to add comment), group *Style*
1. Predefined comments, group *Bug*
1. Predefined comments, group *Unclear*
1. Type in a text comment
1. Record  voice comment
1. Mark of the line that has comments
1. Currently selected line

The *predefined comments* are commonly used comments like *Unformatted code* or *Unused variable*. They can be configured by reviewer so he can add remarks he use most often and then just click the button and select one of the suggestions that appear during the review.

On the long-press line, te screen was divided horizontally so the source code was on the top and the comments added to the line were on the bottom.

I should provide the screenshots of the second app shortly.
